src/upload.rs

```diff
@@ -0,0 +1,205 @@
+const CHUNK_SIZE: u64 = 1024 * 1024 * 10;
+
+async fn do_upload(
+    client: &reqwest::Client,
+    body: impl AsyncRead + Send + 'static,
+    upload_url: &Url,
+    content_length: u64,
+    content_type: Option<&str>,
+    pb: &ProgressBar,
+) -> Result<Response> {
+    let mut request = client
+        .put(upload_url.to_string())
+        .header(AUTHORIZATION, "")
+        .header(CONTENT_LENGTH, content_length);
+    if let Some(content_type) = content_type {
+        request = request.header(CONTENT_TYPE, content_type);
+    }
+    let pb = pb.clone();
+    let body = Body::wrap_stream(ReaderStream::new(body).inspect(move |chunk| {
+        if let Ok(chunk) = chunk.as_ref() {
+            pb.inc(chunk.len() as u64);
+        }
+    }));
+    let response = request.body(body).send().await?;
+    if !response.status().is_success() {
+        Err(error::system(
+            &format!(
+                "Could not upload data: [{}] {}",
+                response.status(),
+                response.text().await.unwrap_or("".to_string())
+            ),
+            "",
+        ))
+    } else {
+        Ok(response)
+    }
+}
+
+async fn simple_upload(
+    client: &reqwest::Client,
+    file: File,
+    upload_url: &Url,
+    content_length: u64,
+    content_type: Option<&str>,
+    pb: &ProgressBar,
+) -> Result<()> {
+    let _guard = TempProgressStyle::new(pb);
+    pb.reset();
+    pb.set_style(progress_bar::pretty_bytes());
+    pb.disable_steady_tick();
+    pb.set_position(0);
+    pb.set_length(content_length);
+    let _ = do_upload(client, file, upload_url, content_length, content_type, pb).await?;
+    Ok(())
+}
+
+async fn upload_part(
+    client: &reqwest::Client,
+    path: impl AsRef<Path>,
+    chunk_number: u64,
+    content_length: u64,
+    upload_url: Url,
+    pb: &ProgressBar,
+) -> Result<String> {
+    let mut file = tokio::fs::File::open(path).await?;
+    file.seek(SeekFrom::Start(chunk_number * CHUNK_SIZE))
+        .await?;
+    let chunk = file.take(CHUNK_SIZE);
+    let response = do_upload(client, chunk, &upload_url, content_length, None, pb).await?;
+    Ok(response
+        .headers()
+        .get("ETag")
+        .ok_or_else(|| error::system("ETag header not found in response", ""))?
+        .to_str()
+        .map_err(|_| error::system("ETag header is not valid UTF-8", ""))?
+        .to_string())
+}
+
+async fn multipart_upload(
+    client: &GraphQLClient,
+    path: impl AsRef<Path>,
+    id: &Id,
+    content_length: u64,
+    content_type: Option<&str>,
+    pb: &ProgressBar,
+) -> Result<()> {
+    let mut chunks = [CHUNK_SIZE].repeat((content_length / CHUNK_SIZE) as usize);
+    if content_length % CHUNK_SIZE != 0 {
+        chunks.push(content_length % CHUNK_SIZE);
+    }
+    let create_multipart_upload = client
+        .send::<CreateMultipartUpload>(create_multipart_upload::Variables {
+            id: id.to_node_id(),
+            chunks: chunks.iter().map(|&x| x as i64).collect(),
```

Comment:
What would be the advantage compared to

```suggestion
            chunk_size: CHUNK_SIZE,
            content_length,
```

?

Reply:
this way the frontend decide how to chunk the data. here its straight forward, but they may have a different chunking strategy where the chunks are not all the same length for some reason. Also its clear what the chunk sizes are. Here i've implemented chunks CHUNK_SIZE * (length / CHUNK_SIZE) + length % CHUNK_SIZE whereas you could also do equal sized chunks given (length / CHUNK_SIZE + 1) chunks or something similar